### PR TITLE
[cli][config-plugins] introduce patch based config-plugin

### PR DESCRIPTION
**Pull request by Kudo**
# Why

close ENG-9909

https://github.com/expo/expo/assets/46429/7baa1b5d-fbd1-4bc5-9193-29ab1aae7d2a

# How

- [cli] introduce `npx expo prepuild:patch` that will convert bare projects to CNG projects and generate **cng-patches/ios.patch+{version}** or **cng-patches/android+{version}.patch**
  - it uses `git diff` to generate patches, so git is required. (patch-package also requires git to generate patches)
  - it moves existing android / ios directories to temporary paths and runs prebuild again. then we can diff between the two project states.
- [config-plugins][prebuild-config] introduce withPatchMod that will apply patches from **cng-patches/ios+{version}.patch** or **cng-patches/android+{version}.patch**
  - the mod will only run once and be the last mod to execute
  - it uses `git apply` to apply the patches, so git is also required in prebuild process if cng-patches files are there. (unlike patch-package that implements patch appliers by their own)

# Test Plan

- added unit tests
- added e2e tests 

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/bin/cli.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ const commands: { [command: string]: () => Promise<Command> } = {
   'run:android': () => import('../src/run/android/index.js').then((i) => i.expoRunAndroid),
   start: () => import('../src/start/index.js').then((i) => i.expoStart),
   prebuild: () => import('../src/prebuild/index.js').then((i) => i.expoPrebuild),
+  uneject: () => import('../src/uneject/index.js').then((i) => i.expoUneject),
```
Comment (brentvatne) [inline packages/@expo/cli/bin/cli.ts:26]: as discussed on slack, let's rename to `prebuild:patch`

File: packages/@expo/cli/src/uneject/index.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      // Aliases
+      '-h': '--help',
+      '-p': '--platform',
+    },
+    argv
+  );
+
+  if (args['--help']) {
+    printHelp(
+      `Convert native iOS and Android project files into CNG patch files`,
```
Comment (brentvatne) [inline packages/@expo/cli/src/uneject/index.ts:24]:
```suggestion
       `(Experimental) Generate patch files for iOS and Android native projects to persist changes made manually after prebuild`
```

**Review by brentvatne — CHANGES_REQUESTED** [review]

**Comment by Kudo** [comment]
some key updates
- change `uneject` to `prebuild:patch` and will keep android/ios folders by default. having a `prebuild:patch --clean` to clean up the folders
- will use prebuild template tarball's md5 checksum as patch version, e.g. `/app/cng-patches/ios+86fb269d190d2c85f6e0468ceca42a20.patch`
  - will show a warning if there are multiple patches
  - will show a warning if patch does not match the prebuild template version. in this case we use the first patch as best effor
- will show a warning if the patch added/deleted lines exceeds 300

now it's good and ready for another round of review

File: packages/@expo/config-plugins/src/plugins/mod-compiler.ts [hunk 3]
```diff
@@ -103,13 +107,15 @@ function getRawClone({ mods, ...config }: ExportedConfig) {
...
+  ios: {
     // dangerous runs first
-    'dangerous',
+    dangerous: 2,
     // run the XcodeProject mod second because many plugins attempt to read from it.
-    'xcodeproj',
-  ],
+    xcodeproj: 1,
+    // patch at the last
+    patch: -1,
```
Comment (EvanBacon) [inline packages/@expo/config-plugins/src/plugins/mod-compiler.ts:117]: Change this to something like `finalize` or `post` to indicate that it runs last. Split this feature out of this PR and land it as a new dangerous modifier that runs after all other modifiers.

```suggestion
    finalize: -1,
```

**Review by EvanBacon — CHANGES_REQUESTED** [review]
The parts of this PR related to adding the new dangerous mod that runs last should be easily mergable in a different PR. The rest of this is a bit shaky.

For the patches that relate to serializable modifiers, I think these should mostly be implemented in some other capacity. Perhaps with a helper plugin that merges static contents in an opinionated way with a simple API.

The pbxproj modifier, which works very different to the other "patches", seems like an entirely different feature. 

I think it's fine if the results of three or four distinct features result in a "patching" system, but we should safely land the more stable concepts first.

**Comment by Kudo** [comment]
i'm closing this pr and split changes to #26413, #26414 and #26429. since the feature is experimental, i'm introducing a new `@expo/prebuild-patch` package to run the `prebuild-patch` command and the config-plugin to apply the patches.
the downside is that requires to `require()` for expo-cli's internal code.

**Pull request closed without merging**